### ISSUE TITLE
[Examples] Add More Training Examples

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2650]
## Problem

Limited examples make it harder for users to learn the library and get started quickly.

### 1. MNIST Digit Classification
```
examples/mnist/
├── train.mojo              # Training script
├── model.mojo              # Simple CNN model
├── README.md               # Setup and results
└── checkpoints/            # Saved weights
```

**Features**:
- Simple CNN architecture (2 conv layers + FC)
- ~99% accuracy target
- Fast training (~5 minutes on CPU)
- Good starter example

### 2. CIFAR-10 Image Classification
```
examples/cifar10/
├── train.mojo              # Training with data augmentation
├── models/
│   ├── resnet.mojo        # ResNet implementation
│   └── vgg.mojo           # VGG implementation
├── data.mojo              # Data loading and preprocessing
└── README.md
```

**Features**:
- More complex architectures
- Data augmentation examples
- Learning rate scheduling
- ~90%+ accuracy target

### 3. Sentiment Analysis (NLP)
```
examples/sentiment/
├── train.mojo              # Text classification
├── model.mojo              # LSTM/GRU model
├── tokenizer.mojo          # Text preprocessing
└── README.md
```

**Features**:
- NLP example
- Sequence processing
- Embedding layers
- RNN/LSTM usage

### 4. GAN (Generative Model)
```
examples/gan/
├── train.mojo              # GAN training
├── generator.mojo          # Generator network
├── discriminator.mojo      # Discriminator network
└── README.md
```

**Features**:
- Adversarial training
- Two-network optimization
- Image generation
- Advanced technique

### 5. Transfer Learning
```
examples/transfer_learning/
├── finetune.mojo           # Fine-tuning pretrained model
├── feature_extraction.mojo # Using as feature extractor
└── README.md
```

**Features**:
- Loading pretrained weights
- Fine-tuning strategies
- Freeze/unfreeze layers
- Practical application

## Common Structure

Each example should include:
- **README.md**: Setup, usage, expected results
- **model.mojo**: Model architecture
- **train.mojo**: Training script
- **infer.mojo**: Inference/evaluation script
- **requirements.txt**: Dependencies (if any)
- **checkpoints/**: Pre-trained weights (optional)

## Benefits

- Learn by example
- Test different architectures
- Benchmark performance
- Demonstrate best practices

## Impact

- **Severity**: LOW
- **Learning**: Easier to get started
- **Adoption**: More complete library

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have gathered enough information. Let me create a comprehensive implementation plan for this issue.

---

# Implementation Plan: Add More Training Examples (Issue #2650)

## Executive Summary

This plan outlines the implementation of 5 new training examples for ML Odyssey to help users learn the library and get started quickly. Based on codebase analysis, **only 1 model (LeNet-5 on EMNIST) is currently fully functional** with complete forward/backward passes. Other models (AlexNet, ResNet, VGG, etc.) have forward-only implementations and are blocked by missing backward passes (especially `batch_norm2d_backward`).

**Key Finding**: The existing infrastructure is mature with excellent patterns to follow (LeNet-5), but several examples proposed in the issue require components that don't yet exist (LSTM/GRU for sentiment analysis, Generator/Discriminator for GAN).

## Analysis of Proposed Examples

| Example | Feasibility | Blockers |
|---------|-------------|----------|
| **1. MNIST Classification** | ✅ HIGH | None - can use LeNet-5 patterns |
| **2. CIFAR-10 Classification** | ⚠️ MEDIUM | Existing ResNet/VGG lack backward passes |
| **3. Sentiment Analysis (NLP)** | ❌ LOW | No LSTM/GRU/Embedding layers |
| **4. GAN (Generative)** | ❌ LOW | Complex training loop, no BatchNorm backward |
| **5. Transfer Learning** | ❌ LOW | Requires pretrained weights, freeze/unfreeze |

## Recommended Phased Approach

Given the current state, I recommend a **phased approach** that delivers value incrementally:

### Phase 1: Immediately Achievable (No New Infrastructure)

1. **MNIST Digit Classification** - Simple CNN using existing ops
2. **CIFAR-10 Simple CNN** - Using only conv/pool/linear (no BatchNorm)

### Phase 2: After Infrastructure Updates

3. **CIFAR-10 with ResNet/VGG** - Requires `batch_norm2d_backward`
4. **Transfer Learning** - Requires pretrained weight support

### Phase 3: Future (Significant Infrastructure)

5. **Sentiment Analysis** - Requires LSTM/GRU/Embedding
6. **GAN** - Requires two-network training loop

---

## Detailed Implementation Plan

### Phase 1, Example 1: MNIST Digit Classification

**Goal**: Create a "Hello World" example simpler than EMNIST (10 classes vs 47)

#### Directory Structure

```text
examples/mnist/
├── model.mojo              # Simple CNN (2 conv + 2 FC)
├── train.mojo              # Training script with manual backward
├── inference.mojo          # Inference script
├── README.md               # Documentation
└── checkpoints/            # Directory for saved weights
```

#### Implementation Tasks

**Task 1.1: Create model.mojo** (~150 lines)

```mojo
# Architecture constants (following LeNet-5 pattern)
alias INPUT_HEIGHT = 28
alias INPUT_WIDTH = 28
alias INPUT_CHANNELS = 1
alias CONV1_OUT_CHANNELS = 32
alias CONV1_KERNEL_SIZE = 3
alias CONV2_OUT_CHANNELS = 64
alias CONV2_KERNEL_SIZE = 3
alias FC1_OUT_FEATURES = 128
alias NUM_CLASSES = 10

struct SimpleCNN(Model, Movable):
    var conv1_kernel: ExTensor
    var conv1_bias: ExTensor
    var conv2_kernel: ExTensor
    var conv2_bias: ExTensor
    var fc1_weights: ExTensor
    var fc1_bias: ExTensor
    var fc2_weights: ExTensor
    var fc2_bias: ExTensor
    
    fn __init__(out self) raises: ...
    fn forward(mut self, input: ExTensor) raises -> ExTensor: ...
    fn predict(mut self, input: ExTensor) raises -> Int: ...
    fn save_weights(self, weights_dir: String) raises: ...
    fn load_weights(mut self, weights_dir: String) raises: ...
    fn update_parameters(mut self, lr: Float32, ...) raises: ...
```

**Task 1.2: Create train.mojo** (~300 lines)

- Parse command-line args with `create_training_parser()`
- Load EMNIST Digits split (simpler than Balanced)
- Implement `compute_gradients()` with manual forward/backward
- Use `TrainingLoop` abstraction
- Evaluate with `evaluate_model_simple()`
- Save weights after training

**Task 1.3: Create inference.mojo** (~150 lines)

- Load saved weights
- Batch evaluation on test set
- Print accuracy and sample predictions

**Task 1.4: Create README.md** (~150 lines)

- Quick start guide
- Architecture diagram
- Expected results (~99% accuracy)
- Training/inference commands

**Task 1.5: Add MNIST download support**

- Extend `scripts/download_emnist.py` to support MNIST variant
- Or create `scripts/download_mnist.py`

#### Files to Modify

1. `examples/mnist/model.mojo` (NEW)
2. `examples/mnist/train.mojo` (NEW)
3. `examples/mnist/inference.mojo` (NEW)
4. `examples/mnist/README.md` (NEW)
5. `scripts/download_emnist.py` (MODIFY - add MNIST split support)
6. `examples/README.md` (MODIFY - add MNIST to directory structure)

#### Success Criteria

- [ ] Model compiles without errors
- [ ] Training completes 10 epochs without crash
- [ ] Achieves ≥98% test accuracy
- [ ] Weights save/load correctly
- [ ] Documentation is complete and accurate
- [ ] Training time < 5 minutes on CPU

---

### Phase 1, Example 2: CIFAR-10 Simple CNN

**Goal**: CIFAR-10 classification without BatchNorm (to avoid backward pass blocker)

#### Directory Structure

```text
examples/cifar10-simple/
├── model.mojo              # Simple CNN without BatchNorm
├── train.mojo              # Training with data augmentation concepts
├── inference.mojo          # Inference script
├── data.mojo               # Data loading helpers
├── README.md               # Documentation
└── checkpoints/
```

#### Implementation Tasks

**Task 2.1: Create model.mojo** (~250 lines)

```mojo
# Simple CNN for CIFAR-10 (no BatchNorm to avoid backward pass blocker)
# Architecture: 
#   Conv(3->32, 3x3) -> ReLU -> MaxPool
#   Conv(32->64, 3x3) -> ReLU -> MaxPool  
#   Conv(64->128, 3x3) -> ReLU -> MaxPool
#   Flatten -> FC(512) -> ReLU -> FC(10)

alias INPUT_HEIGHT = 32
alias INPUT_WIDTH = 32
alias INPUT_CHANNELS = 3  # RGB
alias NUM_CLASSES = 10

struct CIFAR10CNN(Model, Movable):
    # 3 conv layers + 2 FC layers
    var conv1_kernel: ExTensor
    var conv1_bias: ExTensor
    var conv2_kernel: ExTensor
    var conv2_bias: ExTensor
    var conv3_kernel: ExTensor
    var conv3_bias: ExTensor
    var fc1_weights: ExTensor
    var fc1_bias: ExTensor
    var fc2_weights: ExTensor
    var fc2_bias: ExTensor
```

**Task 2.2: Create train.mojo** (~400 lines)

- Use existing `scripts/download_cifar10.py`
- Load CIFAR-10 using `shared/data/datasets/cifar10.mojo`
- Implement manual forward/backward pass
- Add learning rate scheduling (step decay)
- Target: ~70-75% accuracy (without BatchNorm/augmentation)

**Task 2.3: Create data.mojo** (~100 lines)

- Wrapper functions for CIFAR-10 loading
- Normalization (mean subtraction)
- One-hot encoding

**Task 2.4: Create inference.mojo** (~150 lines)

- Load weights
- Evaluate on test set
- Show per-class accuracy

**Task 2.5: Create README.md** (~200 lines)

- Architecture explanation
- Training tips
- Expected results
- Comparison with more complex architectures

#### Files to Modify/Create

1. `examples/cifar10-simple/model.mojo` (NEW)
2. `examples/cifar10-simple/train.mojo` (NEW)
3. `examples/cifar10-simple/data.mojo` (NEW)
4. `examples/cifar10-simple/inference.mojo` (NEW)
5. `examples/cifar10-simple/README.md` (NEW)
6. `examples/README.md` (MODIFY)

#### Success Criteria

- [ ] Model compiles without errors
- [ ] Training completes 50 epochs without crash
- [ ] Achieves ≥70% test accuracy
- [ ] Demonstrates learning rate scheduling
- [ ] Documentation explains BatchNorm limitation

---

### Phase 2: Infrastructure-Dependent Examples

#### Phase 2, Example 3: CIFAR-10 with ResNet (Pending Issue #2576)

**Blocker**: `batch_norm2d_backward` implementation

**Prerequisites**:

1. Implement `batch_norm2d_backward` in `shared/core/normalization.mojo`
2. Complete ResNet-18 backward pass in `examples/resnet18-cifar10/train.mojo`

**Files to Create/Modify**:

1. `shared/core/normalization.mojo` - Add `batch_norm2d_backward`
2. `examples/resnet18-cifar10/train.mojo` - Complete backward pass
3. `examples/cifar10/models/resnet.mojo` - Wrapper model
4. `examples/cifar10/README.md` - Update documentation

#### Phase 2, Example 4: Transfer Learning

**Blocker**: Pretrained weight loading, freeze/unfreeze layer support

**Prerequisites**:

1. Implement layer freezing (skip gradient computation)
2. Create pretrained weight repository
3. Add weight download utilities

**Files to Create**:

1. `examples/transfer_learning/finetune.mojo`
2. `examples/transfer_learning/feature_extraction.mojo`
3. `examples/transfer_learning/README.md`
4. `shared/training/freeze_utils.mojo` - Freeze/unfreeze helpers
5. `scripts/download_pretrained.py` - Pretrained weight downloader

---

### Phase 3: Future Examples (Significant Infrastructure Required)

#### Phase 3, Example 5: Sentiment Analysis (NLP)

**Blockers**: No LSTM/GRU/Embedding layers

**Required New Components**:

1. `shared/core/embedding.mojo` - Embedding layer with lookup
2. `shared/core/rnn.mojo` - LSTM/GRU cells with forward/backward
3. `shared/data/text_tokenizer.mojo` - Full tokenizer (current text_transforms.mojo is basic)
4. Dataset loaders (IMDB, SST-2)

**Estimated Effort**: 2-3 weeks

#### Phase 3, Example 6: GAN (Generative Adversarial Network)

**Blockers**: Complex training loop, BatchNorm backward

**Required New Components**:

1. `batch_norm2d_backward` (same as Phase 2)
2. Two-network training loop abstraction
3. Generator/Discriminator loss functions
4. Image generation utilities

**Estimated Effort**: 2-3 weeks after Phase 2

---

## Testing Approach

### Unit Tests (for each example)

Create `tests/examples/test_<example>_model.mojo`:

```mojo
# Test model forward pass shape
fn test_forward_shape() raises:
    var model = SimpleCNN()
    var input = zeros([1, 1, 28, 28], DType.float32)
    var output = model.forward(input)
    assert_equal(output.shape(), [1, 10])

# Test gradient flow
fn test_gradient_computation() raises:
    var model = SimpleCNN()
    var input = special_test_values([2, 1, 28, 28])  # Use FP-safe values
    var labels = one_hot([0, 1], 10)
    var loss = compute_gradients(model, input, labels, 0.01)
    assert_true(loss > 0.0)

# Test weight save/load
fn test_weight_serialization() raises:
    var model = SimpleCNN()
    model.save_weights("/tmp/test_weights")
    var loaded = SimpleCNN()
    loaded.load_weights("/tmp/test_weights")
    # Compare parameters
```

### Integration Tests

- Run training for 1 epoch, verify loss decreases
- Run inference on saved weights, verify accuracy > random

### CI Integration

Add to `.github/workflows/comprehensive-tests.yml`:

```yaml
- name: Test MNIST example
  run: just ci-test-group "tests/examples" "test_mnist*.mojo"
```

---

## Common Structure for All Examples

Each example should include:

```text
examples/<name>/
├── README.md               # Setup, usage, expected results
├── model.mojo              # Model architecture with:
│   ├── Architecture constants (alias)
│   ├── Struct implementing Model trait
│   ├── forward(), predict()
│   ├── save_weights(), load_weights()
│   └── update_parameters()
├── train.mojo              # Training script with:
│   ├── TrainConfig struct
│   ├── parse_args() using create_training_parser()
│   ├── compute_gradients() with manual forward/backward
│   ├── train_epoch() using TrainingLoop
│   └── main() orchestrating flow
├── inference.mojo          # Inference script with:
│   ├── Weight loading
│   ├── Batch evaluation
│   └── Metrics reporting
└── checkpoints/            # Directory for saved weights
```

---

## Dependency Graph

```text
Phase 1 (No Dependencies)
├── MNIST Example
│   └── Depends on: LeNet-5 patterns (existing)
└── CIFAR-10 Simple
    └── Depends on: LeNet-5 patterns (existing)

Phase 2 (Infrastructure Dependencies)
├── CIFAR-10 ResNet
│   └── Depends on: batch_norm2d_backward (Issue #2576)
└── Transfer Learning
    └── Depends on: Freeze/unfreeze support (new)

Phase 3 (Major Infrastructure)
├── Sentiment Analysis
│   └── Depends on: LSTM/GRU/Embedding (new)
└── GAN
    └── Depends on: batch_norm2d_backward + training loop (new)
```

---

## Estimated Timeline

| Phase | Example | Effort | Prerequisites |
|-------|---------|--------|---------------|
| 1 | MNIST | 2-3 days | None |
| 1 | CIFAR-10 Simple | 3-4 days | None |
| 2 | CIFAR-10 ResNet | 1 week | batch_norm2d_backward |
| 2 | Transfer Learning | 1 week | Freeze support |
| 3 | Sentiment Analysis | 2-3 weeks | LSTM/GRU/Embedding |
| 3 | GAN | 2-3 weeks | Phase 2 complete |

---

## Implementation Priority

**Recommended order for Phase 1**:

1. **MNIST Example** (highest impact, easiest to implement)
   - Simpler than EMNIST (10 classes vs 47)
   - Perfect "Hello World" for new users
   - Builds confidence before CIFAR-10

2. **CIFAR-10 Simple** (second priority)
   - RGB images (more realistic)
   - No BatchNorm blocker
   - Foundation for Phase 2 examples

---

## Files Summary

### New Files to Create

| File | Phase | Lines (est.) |
|------|-------|--------------|
| `examples/mnist/model.mojo` | 1 | ~150 |
| `examples/mnist/train.mojo` | 1 | ~300 |
| `examples/mnist/inference.mojo` | 1 | ~150 |
| `examples/mnist/README.md` | 1 | ~150 |
| `examples/cifar10-simple/model.mojo` | 1 | ~250 |
| `examples/cifar10-simple/train.mojo` | 1 | ~400 |
| `examples/cifar10-simple/data.mojo` | 1 | ~100 |
| `examples/cifar10-simple/inference.mojo` | 1 | ~150 |
| `examples/cifar10-simple/README.md` | 1 | ~200 |
| `tests/examples/test_mnist_model.mojo` | 1 | ~100 |
| `tests/examples/test_cifar10_simple.mojo` | 1 | ~100 |

### Files to Modify

| File | Changes |
|------|---------|
| `examples/README.md` | Add new examples to directory structure |
| `scripts/download_emnist.py` | Add MNIST split support |
| `.github/workflows/comprehensive-tests.yml` | Add example tests |

---

## Success Criteria (Overall)

### Phase 1 Complete When

- [ ] MNIST example achieves ≥98% accuracy
- [ ] CIFAR-10 Simple achieves ≥70% accuracy
- [ ] Both examples have complete documentation
- [ ] Both examples have passing unit tests
- [ ] Training times are reasonable (< 30 min CPU)
- [ ] Code follows existing patterns (LeNet-5)

### Phase 2 Complete When

- [ ] `batch_norm2d_backward` is implemented
- [ ] ResNet-18 full training works
- [ ] Transfer learning example demonstrates freeze/unfreeze
- [ ] CIFAR-10 with ResNet achieves ≥90% accuracy

### Phase 3 Complete When

- [ ] LSTM/GRU/Embedding layers implemented
- [ ] Sentiment analysis achieves ≥85% accuracy
- [ ] GAN generates recognizable digits
- [ ] All 5 proposed examples functional

---

## Risks and Mitigations

| Risk | Impact | Mitigation |
|------|--------|------------|
| BatchNorm backward complexity | Blocks Phase 2 | Focus on Phase 1 first |
| LSTM/GRU implementation time | Blocks Phase 3 | Keep as future work |
| CIFAR-10 accuracy without BatchNorm | Lower accuracy | Set realistic expectations (~70%) |
| Training time on CPU | Poor UX | Use small architectures, add GPU support notes |

---

## Resource Usage

- Tool calls used: 12/50
- Steps used: 7/50
